modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/action/BulkCopyEntriesMVCActionCommand.java

```diff
@@ -0,0 +1,174 @@
+			ActionRequest actionRequest, ActionResponse actionResponse)
+		throws Exception {
+
+		ThemeDisplay themeDisplay = (ThemeDisplay)actionRequest.getAttribute(
+			WebKeys.THEME_DISPLAY);
+
+		try {
+			_copyEntities(actionRequest);
+		}
+		catch (PortalException portalException) {
+			String errorMessage = StringBundler.concat(
+				portalException.getMessage(), StringPool.SPACE,
+				themeDisplay.translate(
+					"documents-folders-could-not-be-copied"));
+
+			if (_errorsMap.size() < 10) {
+				errorMessage = StringBundler.concat(
+					portalException.getMessage(), StringPool.SPACE,
+					themeDisplay.translate(
+						"the-following-documents-folders-could-not-be-copied"));
+			}
+
+			JSONPortletResponseUtil.writeJSON(
+				actionRequest, actionResponse,
+				JSONUtil.put("errorMessage", errorMessage));
+
+			hideDefaultSuccessMessage(actionRequest);
+
+			_errorsMap.clear();
+		}
+	}
+
+	private void _checkDestinationGroup(Group group) throws PortalException {
+		if ((group != null) && group.isStaged() && !group.isStagingGroup()) {
+			throw new PortalException(
+				"cannot-copy-into-the-live-version-of-a-group");
+		}
+	}
+
+	private void _copyEntities(ActionRequest actionRequest)
+		throws PortalException {
+
+		long destinationFolderId = ParamUtil.getLong(
+			actionRequest, "destinationFolderId");
+		long destinationRepositoryId = ParamUtil.getLong(
+			actionRequest, "destinationRepositoryId");
+		long[] entries = ParamUtil.getLongValues(
+			actionRequest, "selectedEntries");
+
+		Group group = _groupLocalService.fetchGroup(destinationRepositoryId);
+
+		_checkDestinationGroup(group);
+
+		long[] currentAndAncestorSiteAndDepotGroupIds =
+			_siteConnectedGroupGroupProvider.
+				getCurrentAndAncestorSiteAndDepotGroupIds(group.getGroupId());
+
+		for (long entryId : entries) {
+			DLFileShortcut dlFileShortcut =
+				_dlFileShortcutLocalService.fetchDLFileShortcut(entryId);
+
+			DLFolder dlFolder = _dlFolderLocalService.fetchDLFolder(entryId);
+
+			try {
+				if (_dlFileEntryLocalService.fetchDLFileEntry(entryId) !=
+						null) {
+
+					_dlAppService.copyFileEntry(
+						entryId, destinationFolderId, destinationRepositoryId,
+						currentAndAncestorSiteAndDepotGroupIds,
+						ServiceContextFactory.getInstance(
+							DLFileEntry.class.getName(), actionRequest));
+				}
+				else if (dlFileShortcut != null) {
+					_dlAppService.copyFileShortcut(
+						entryId, destinationFolderId, destinationRepositoryId,
+						ServiceContextFactory.getInstance(
+							DLFileShortcut.class.getName(), actionRequest));
+				}
+				else if (dlFolder != null) {
+					_dlAppService.copyFolder(
+						dlFolder.getRepositoryId(), entryId,
+						destinationRepositoryId, destinationFolderId,
+						currentAndAncestorSiteAndDepotGroupIds,
+						ServiceContextFactory.getInstance(
+							DLFolder.class.getName(), actionRequest));
+				}
+			}
+			catch (PortalException portalException) {
+				_errorsMap.put(entryId, portalException.getMessage());
```

Comment:
we need more information about the message/file failing.

instead of the portal exception we need the name or something similar, no?